garden_ai/app/pipeline.py

```diff
@@ -565,11 +568,108 @@ def container(
+        resolve_path=True,
+        help=("Path to a Python file containing your pipeline implementation."),
+    ),
+    docker: bool = typer.Option(
+        False,
+        "--docker",
+        help="When set, the pipeline session will be loaded within an automatically built container",
+    ),
+    cleanup: bool = typer.Option(
+        False,
+        "--rm",
+        help="When set, the built container image will be removed automatically (only used when --docker is also supplied)",
+    ),
+):
+    import subprocess
+    from functools import partial
+
+    client = GardenClient()
+
+    if (
+        not pipeline_file.exists()
+        or not pipeline_file.is_file()
+        or pipeline_file.suffix != ".py"
+    ):
+        console.log(
+            f"{pipeline_file} is not a valid Python file. Please provide a valid Python file (.py)."
+        )
+        raise typer.Exit(code=1)
+    try:
+        user_pipeline = load_pipeline_from_python_file(pipeline_file)
+        pipeline_url_json = client.generate_presigned_urls_for_pipeline(user_pipeline)
+        _env_vars = {GardenConstants.URL_ENV_VAR_NAME: pipeline_url_json}
+    except (
+        PipelineLoadException,
+        PipelineLoadScaffoldedException,
+    ) as e:
+        console.log(f"Could not parse {pipeline_file} as a Garden pipeline. " + str(e))
+        raise
+    remote_func = make_func_to_serialize(user_pipeline)
+    curry = partial(remote_func, _env_vars=_env_vars)
+
+    with TemporaryDirectory() as tmpdir:
+        tmpdir = tmpdir.replace("\\", "/")  # Windows compatibility
+        with NamedTemporaryFile(dir=tmpdir, delete=False) as tmp:
+            dill.dump(curry, tmp)
+            tmp_filename = Path(tmp.name).name
+
+        # contains everything, control flow dictates which pieces are removed
+        interpreter_cmd = [
+            "python",
+            "-i",
+            "-c",
+            f'\'import dill; f = open("{tmpdir}/{tmp_filename}", "rb"); {user_pipeline.short_name} = dill.load(f); f.close(); del f; '
+            f'print(f"Defined your pipeline as the function `{user_pipeline.short_name}`!")\'',
+        ]
+
+        if docker:
+            image_name = f"gardenai/{user_pipeline.short_name}"
+            # avoids SyntaxError in inline version
+            escaped_pip_deps = [f'"{dep}"' for dep in user_pipeline.pip_dependencies]
+            # remove tmpdir in container's path
+            interpreter_cmd[-1] = interpreter_cmd[-1].replace(f"{tmpdir}/", "")
+
+            with open(f"{tmpdir}/Dockerfile", "w+") as dockerfile:
+                dockerfile.write(
+                    "FROM continuumio/miniconda3\n"
+                    f'ARG SERIAL="{tmp_filename}"\n'  # allows filepaths with spaces (precaution)
+                    "COPY $SERIAL .\n"
+                    f"RUN conda create -y -n env pip python={user_pipeline.python_version}\n"
+                    'SHELL ["conda", "run", "--no-capture-output", "-n", "env", "/bin/bash", "-c"]\n'
```

Comment:
All the individual phrases of code in this function look solid to me. But at this point in the function we're getting close to 100 lines and we're quite indented. (Nothing wrong with indentation as such, just a convenient rule of thumb for looking at complexity of a function.) Are there any natural places where you might break up this function?

That said, < 100 lines isn't egregious and this is kind of experimental. So I'm throwing out the suggestion to break this up a bit but up to you.